usb: device_next: add new MIDI 2.0 device class

titouanc opened the pull request:
Add implementation for the 2nd revision of the USB MIDI device class (the MIDIStreaming subclass of the Audio device class), based on the new USB device stack. Moreover, add a sample application to demonstrate usage of this new device class.

This implementation is based on the following documents:

* `midi20`: [Universal Serial Bus Device Class Definition for MIDI Devices Release 2.0](https://www.usb.org/sites/default/files/USB%20MIDI%20v2_0.pdf)
* `ump112`: [Universal MIDI Packet (UMP) Format and MIDI 2.0 Protocol With MIDI 1.0 Protocol in UMP Format _Document Version 1.1.2_](https://midi.org/universal-midi-packet-ump-and-midi-2-0-protocol-specification)

In this initial implementation, the device class only supports USB-MIDI2.0 (which can convey MIDI1 and MIDI2 data). However, the USB-MIDI2.0 specification requires to define a valid USB-MIDI1.0 interface before the 2.0 one for backward compatibility. Therefore a single interface defined in the device tree will yield 2 USB interfaces (with altsetting 0 and 1). The first one is always a "dummy" one (the minimal interface without any input/output), and the inputs/outputs of the second one are the ones defined in the device tree. As such, data exchange is only possible if the MIDI2.0 (altsetting 1) has been enabled by the host.

Zephyr application code can exchange MIDI data with the host in the form of Universal MIDI Packets through "Group Terminals", as illustrated in `midi20: 4. Operational model`:
![image](https://github.com/user-attachments/assets/94aff24d-2317-48db-a19d-2be53d68238b)

titouanc committed:
usb: device_next: add new MIDI 2.0 device class

This adds a new USB device class (based on usb/device_next) that implements
revision 2.0 of the MIDIStreaming interface, a sub-class of the USB audio
device class. In practice, the MIDI interface is much more simple and has
little in common with Audio, so it makes sense to have it as a separate
class driver.

MIDI inputs and outputs are configured through the device tree, under a
node `compatible = "zephyr,usb-midi"`. As per the USB-MIDI2.0 spec,
a single usb-midi interface can convey up to 16 Universal MIDI groups,
comprising 16 channels each. Data is carried from/to the host via
so-called Group Terminals, that are organized in Group Terminal Blocks.
They are represented as children of the usb-midi interface in the device
tree.

From the Zephyr application programmer perspective, MIDI data is exchanged
with the host through the device associated with the `zephyr,usb-midi`
interface, using the following API:

* Send a Universal MIDI Packet to the host: `usb_midi_send(device, pkt)`
* Universal MIDI Packets from the host are delivered to the function passed
  in `usb_midi_set_ops(device, &{.rx_packet_cb = handler})`

Compliant USB-MIDI 2.0 devices are required to expose a USB-MIDI1.0
interface as alt setting 0, and the 2.0 interface on alt setting 1.
To avoid the extra complexity of generating backward compatible USB
descriptors and translating Universal MIDI Packets from/to the old
USB-MIDI1.0 format, this driver generates an empty MIDI1.0 interface
(without any input/output); and therefore will only be able to exchange
MIDI data when the host has explicitely enabled MIDI2.0 (alt setting 1).

This implementation is based on the following documents, which are referred
to in the inline comments:

* `midi20`:
    Universal Serial Bus Device Class Definition for MIDI Devices
    Release 2.0
    https://www.usb.org/sites/default/files/USB%20MIDI%20v2_0.pdf
* `ump112`:
    Universal MIDI Packet (UMP) Format and MIDI 2.0 Protocol
      With MIDI 1.0 Protocol in UMP Format
    Document Version 1.1.2
    https://midi.org/universal-midi-packet-ump-and-midi-2-0-protocol-specification

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
samples: subsys: usb: midi: add new sample

Add a sample application that demonstrates how to use the new USB-MIDI 2.0
device class. This shows how to set up the device tree, how to exchange
MIDI data with the host, and how to use the data accessors provided by
the new API.

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,169 @@
+/**
+ * @brief USB-MIDI 2.0 class device API
+ * @defgroup usb_midi USB MIDI 2.0 Class device API
+ * @ingroup usb
```
```suggestion
 * @ingroup usb
 * @since 4.1
 * @version 0.1.0
```

tmon-nordic reviewed (changes requested):
Are you planning on supporting MIDI 1.0 later? The specification strongly recomments, but does not mandate the MIDI 1.0 backwards compatible interface.

Do you have idea about how to support System Exclusive commands? Should "slicing" the SysEx message be application duty or class?

titouanc commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,743 @@
+#define ALT_USB_MIDI_1 0x00
+#define ALT_USB_MIDI_2 0x01
+
+/* audio10 4.3.2 Class-Specific AC Interface Descriptor */
```
I found the definition of this descriptor in "Universal Serial Bus Device Class Definition for Audio Devices - Release 1.0 (1998-03-18)", which is referred to in `midi20: 1.3 Related documents`.

Agreed that this is an unnecessary indirection, will change to `midi20: B.2.2 Class-specific AC Interface Descriptor`

titouanc commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,743 @@
+	switch (alternate) {
+	case ALT_USB_MIDI_1:
+		data->altsetting = ALT_USB_MIDI_1;
+		LOG_WRN("%s set USB-MIDI1.0 altsetting (not implemented !)", dev->name);
```
Only when the host selects altsetting 0 (MIDI1.0)

tmon-nordic replied:
The fact that this is not being called sounds like a bug in the USB stack to me. I believe `usbd_interface_default()` should really call the update callback with alt setting 0 if things behaved like documented:
https://github.com/zephyrproject-rtos/zephyr/blob/db4dc5b450d8518d1999779117e563f25fba58df/subsys/usb/device_next/usbd_class_api.h#L137-L139

Currently the only option to get class update callback called is that host first selects non-zero alternate and then switches back to zero alternate.

Not sure how this should be fixed (options are either to change the `usbd_class_update()` description or to have the `usbd_class_update()` called in `usbd_interface_default()`), but that's definitely to be handled outside this PR.

tmon-nordic replied:
> No, there is no need to change anything

Then why https://github.com/zephyrproject-rtos/zephyr/blob/db4dc5b450d8518d1999779117e563f25fba58df/subsys/usb/device_next/usbd_class_api.h#L137-L139
states "because of Set Configuration"? The Set Configuration alone will never result in `usbd_class_update()`, no matter how many alternate settings there are or what alternate setting was previously selected.

titouanc commented:
Thank you @tmon-nordic for your first comments !

> Are you planning on supporting MIDI 1.0 later? The specification strongly recomments, but does not mandate the MIDI 1.0 backwards compatible interface.

> Do you have idea about how to support System Exclusive commands? Should "slicing" the SysEx message be application duty or class?

I decided to go for USB-MIDI2.0, because it transports Universal MIDI Packets (UMPs). The intended benefit is that this is an externally defined spec, independent of the transport layer, supporting both MIDI1 and MIDI2, and carrying "routing" (MIDI group) metadata. I believe these packets could be readily used on other transports such as Bluetooth or IP (but this goes beyond my knowledge). On the contrary, USB-MIDI1.0 has a packet format that is specific to this transport, and therefore somehow "leaks" some implementation details that are specific to USB-MIDI.

Moreover, I decided to NOT use a stream-based API (`midi_write/read(some_data_bytes)` like a serial port), because it avoids the burden of chunking MIDI (application) data into USB-MIDI packets. Such an implementation would require a MIDI parser and some kind of internal state; which I believe is out of scope of a USB class driver. As you noted, this implies that application code has to handle the extra complexity of splitting up long SysEx commands into valid UMPs. If there's a real need here, we could add the following function to the API, that would be in charge of chunking & sending a larger buffer into individual UMPs:

```C
int usbd_midi_send_sysex(const struct device *dev,
                         const uint8_t *data, size_t size);
```

Overall, sending/receiving simple MIDI events (note on/off, control changes, system messages, short sysex, etc...) is straightforward with this proposal. Long SysEx are already some kind of "advanced" use cases, so I wouldn't try to do more about it here.

Even though USB-MIDI1.0 transport is not supported, I noticed that I needed to add that "dummy" backward compatible interface before the USB-MIDI2.0 one, otherwise the USBD MIDI Sample wouldn't enumerate properly. I tested both with Linux 6.11/alsa 1.2.12 and Mac OS X 15.1, but I'd be glad to hear about any experience with other hosts (like Android, iOS or Windows). For the record, here is how it looks on Mac OS:

![image](https://github.com/user-attachments/assets/80d22cbf-c96a-47a7-95cc-4893c29a449d)

And on Linux:

```
$ amidi -l
Dir Device    Name
IO  hw:2,1,0  Group 1 (USBD MIDI Sample)
```

I haven't planned on implementing a USB-MIDI1.0 stack, because 2.0 works for me, and its features are a superset of 1.0. Moreover, I haven't a clear idea on how it would look like for application code (are the 1.0/2.0 altsetting distinguably usable from dt/code ? How does the application code know which is selected, and how to format packets ? Otherwise do we have a translation layer 1.0<->2.0 depending on the selected altsetting ?).

Finally, I will dig into the full-speed/high-speed descriptors and amend the PR accordingly. 

Thank you again for looking at this !

tmon-nordic commented:
> Even though USB-MIDI1.0 transport is not supported, I noticed that I needed to add that "dummy" backward compatible interface before the USB-MIDI2.0 one, otherwise the USBD MIDI Sample wouldn't enumerate properly.

It wouldn't enumerate properly because it is mandatory for the MIDI 2.0 interface to be on alternate setting 1. Take a look at USB MIDI 2.0 Specification and note where "should" (recommended but optional) is used and where "shall" (mandatory) is used.

> If there's a real need here, we could add the following function to the API, that would be in charge of chunking & sending a larger buffer into individual UMPs

I don't quite know about MIDI 2.0 but in MIDI 1.0 the most significant bit is reserved for Real Time messages. All SysEx commands (and responses) do have the most significant bit cleared to enable having the Real Time message while the SysEx is being transmitted. To what degree this is a problem is unknown to me, but USB-MIDI 1.0 chunks everything into 4 byte USB-MIDI Event Packets that can be freely interleaved.

While this API would be useful for applications that use SysEx messages exclusively (e.g. Digitech effect pedals, see https://github.com/desowin/gdigi for open-source host-side implementation), I have no idea if would introduce problems when chunk interleaving is necessary.

> Moreover, I haven't a clear idea on how it would look like for application code (are the 1.0/2.0 altsetting distinguably usable from dt/code ? How does the application code know which is selected, and how to format packets ? Otherwise do we have a translation layer 1.0<->2.0 depending on the selected altsetting ?).

This is something that would have to be solved if we wanted proper backwards compatibility. The alt setting is only known at runtime and is entirely host-dependent. The class would have to somehow provide the information about selected protocol version to the application. About the format conversion I have no idea.

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,171 @@
+ * @see ump112: 7.3 MIDI 1.0 Channel Voice Messages
+ */
+struct ump_midi1 {
+#ifdef CONFIG_LITTLE_ENDIAN
+	/** @brief The 2nd MIDI1 parameter */
+	uint8_t p2;
+	/** @brief The 1st MIDI1 parameter */
+	uint8_t p1;
+	/** @brief The MIDI1 channel number */
+	unsigned channel: 4;
+	/** @brief The MIDI1 status nibble (command) */
+	unsigned status: 4;
+	/** @brief The UMP group to which this message belongs */
+	unsigned group: 4;
+	/** @brief The UMP message type (MT_MIDI1_CHANNEL_VOICE) */
+	unsigned mt: 4;
+#else
+	/** @brief The UMP message type (MT_MIDI1_CHANNEL_VOICE) */
+	unsigned mt: 4;
+	/** @brief The UMP group to which this message belongs */
+	unsigned group: 4;
+	/** @brief The MIDI1 status nibble (command) */
+	unsigned status: 4;
+	/** @brief The MIDI1 channel number */
+	unsigned channel: 4;
+	/** @brief The 1st MIDI1 parameter */
+	uint8_t p1;
+	/** @brief The 2nd MIDI1 parameter */
+	uint8_t p2;
+#endif
```
I leave it up to you to decide whether to use a macro to set the bitfields on `uint32_t` here or rely on compiler putting things in correct order. The bit order is in no way guaranteed by the standard. I do know that parts of USB stack use this approach, but I don't quite like it.

titouanc replied:
Yes indeed, I don't know what's the best way of defining midi packet fields here. I got some inspiration from other USB structs (for example https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/usb/usb_ch9.h#L27-L49).

While I like being able to access individual fields from the midi packet as a struct, this is tedious and error prone, because all fields have to be repeated in the correct order twice.

Not sure if there is some kind of convention to follow here ? I'd happily change that part

tmon-nordic replied:
I don't know if there is any specific convention for this type of things. The portable code to handle such things is to have macros to put fields on correct bit places. I am not really familiar with actual MIDI usage so cannot really come up with what sort of macros would work best.

titouanc replied:
I just pushed a new version (rebased on the latest main) that replace these enums with accessors/constructor macros on `uint32_t []`

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,171 @@
+union ump {
+	uint32_t words[4];
+	struct {
+#ifdef CONFIG_LITTLE_ENDIAN
+		unsigned _reserved: 24;
+		/** @brief The UMP group to which this message belongs */
+		unsigned group: 4;
+		/** @brief The UMP message type */
+		unsigned mt: 4;
+#else
+		/** @brief The UMP message type */
+		unsigned mt: 4;
+		/** @brief The UMP group to which this message belongs */
+		unsigned group: 4;
+#endif
```
Same as above, but here the big endian part misses the reserved field.

kartben commented:
https://github.com/user-attachments/assets/e928d660-1e88-472e-b0a0-dd0eb7fa91a7

@titouanc  this is really cool :)

kartben commented on samples/subsys/usb/midi/src/main.c:
```diff
@@ -0,0 +1,72 @@
+
+static const struct device *const midi = DEVICE_DT_GET(DT_NODELABEL(usb_midi));
+
+/* On boards that have a user button; button press/release -> MIDI note on/off */
+#if DT_NODE_EXISTS(DT_NODELABEL(user_button))
+static const struct device *const button = DEVICE_DT_GET(DT_PARENT(DT_NODELABEL(user_button)));
+
+static void key_press(struct input_event *evt, void *user_data)
+{
+	uint8_t command = evt->value ? MIDI_NOTE_ON : MIDI_NOTE_OFF;
+
+	usb_midi_send(midi, &UMP_MIDI1(0, command, 0, 0x40, 0x7f));
+}
+INPUT_CALLBACK_DEFINE(button, key_press, NULL);
+#endif
```
sw0 alias is the more common/standard way of describing the "main" button on a board, so switch to that.
In fact, you may possibly look for sw1, sw2, ... too while you're at it, and play a n+1, n+2, etc. pitch when those are hit.

I would register the callback using `INPUT_CALLBACK_DEFINE(NULL, key_press, NULL);` and check keycodes in the callback (you may get the keycodes for sw0, sw1, etc. using DT_PROP macro). In fact, right now the way your callback is registered and written, it actually reacts to any button being pressed on the same gpio controller as "user_button", which is probably not what you want :)

kartben commented on samples/subsys/usb/midi/README.rst:
```diff
@@ -0,0 +1,81 @@
+.. zephyr:code-sample:: midi
+   :name: USB MIDI device class sample
```
```suggestion
   :name: USB MIDI device class
```

kartben commented on samples/subsys/usb/midi/README.rst:
```diff
@@ -0,0 +1,81 @@
+   :name: USB MIDI device class sample
+   :relevant-api: usbd_api usb_midi
+
+   USB MIDI 2.0 device class sample
```
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/templates/sample.tmpl#L5-L6

```suggestion
   Implement a simple USB MIDI 2.0 device that loops back MIDI messages and sends note events using a user button.
```

bonus points if you find a way to make it shorter. "Implement a simple USB MIDI 2.0 device" would probably work, actually.

kartben commented on samples/subsys/usb/midi/README.rst:
```diff
@@ -0,0 +1,81 @@
+  IO  hw:2,1,0  Group 1 (USBD MIDI Sample)
+
+
+The "USBD MIDI Sample" interface should also appear in any program with MIDI
+support (for example your favorite Digital Audio Workstation)
```
maybe also mention some of the web based tools that exist, as this makes things even easier to test. We could *almost* have a minimal hosted html/js page similar to https://docs.zephyrproject.org/latest/samples/subsys/usb/webusb/demo.html in the tree to make it dead easy for folks to test / play with this.

titouanc replied:
I wasn't familiar with webusb so I looked into that, and I discovered that there's even a webmidi API. I would rather find a demo based on the latter, otherwise we'd have to re-build usb-midi host support in the demo :stuck_out_tongue: 

By the way, is your video based on this website https://muted.io/piano/ ? Maybe we could directly refer to that ?

toonst commented on samples/subsys/usb/midi/README.rst:
```diff
@@ -0,0 +1,81 @@
+
+.. code-block:: console
+
+  $ amidi -p hw:2,1,0 -d
```
Alternative to this could be @gbevin's cross platform [sendMidi](https://github.com/gbevin/SendMIDI)

titouanc replied:
As @kartben suggested, we could also use some web-based tool (which I expect would be more portable and easy to distribute)

kartben replied:
FWIW https://muted.io/piano/ is awesome, but it's probably a stretch to mention it in the official Zephyr docs...

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,149 @@
+ * @param[in]  ump    The universal MIDI Packet
+ * @see ump112: 2.1.4 Message Type (MT) Allocation
+ */
+#define UMP_WORDS(ump) 1 + ((0xfe950d40 >> (2 * UMP_MT(ump))) & 3)
```
```suggestion
#define UMP_WORDS(ump) (1 + ((0xfe950d40 >> (2 * UMP_MT(ump))) & 3))
```

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,149 @@
+ * @brief      MIDI group field of a Universal MIDI Packet
+ * @param[in]  ump    The universal MIDI Packet
+ */
+#define UMP_GROUP(ump) ((ump)[0] >> 24) & 0x0f
```
```suggestion
#define UMP_GROUP(ump) (((ump)[0] >> 24) & 0x0f)
```

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,149 @@
+ * @brief      Command of a MIDI1 channel voice message
+ * @param[in]  ump    A universal MIDI Packet (containing a MIDI1 event)
+ */
+#define UMP_MIDI1_COMMAND(ump) ((ump)[0] >> 20) & 0x0f
```
```suggestion
#define UMP_MIDI1_COMMAND(ump) (((ump)[0] >> 20) & 0x0f)
```

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,149 @@
+ * @brief      Channel of a MIDI1 channel voice message
+ * @param[in]  ump    A universal MIDI Packet (containing a MIDI1 event)
+ */
+#define UMP_MIDI1_CHANNEL(ump) ((ump)[0] >> 16) & 0x0f
```
```suggestion
#define UMP_MIDI1_CHANNEL(ump) (((ump)[0] >> 16) & 0x0f)
```

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,149 @@
+ * @brief      First parameter of a MIDI1 channel voice message
+ * @param[in]  ump    A universal MIDI Packet (containing a MIDI1 event)
+ */
+#define UMP_MIDI1_P1(ump)      ((ump)[0] >> 8) & 0x7f
```
```suggestion
#define UMP_MIDI1_P1(ump)      (((ump)[0] >> 8) & 0x7f)
```

tmon-nordic commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,149 @@
+ * @brief      Second parameter of a MIDI1 channel voice message
+ * @param[in]  ump    A universal MIDI Packet (containing a MIDI1 event)
+ */
+#define UMP_MIDI1_P2(ump)      (ump)[0] & 0x7f
```
```suggestion
#define UMP_MIDI1_P2(ump)      ((ump)[0] & 0x7f)
```

tmon-nordic reviewed (commented):
Wrap macro values in parentheses to avoid potential issues on macro use.

titouanc commented:
I was looking for a nice web based demo as you suggested, before finalizing the sample app @kartben

kartben commented:
> I was looking for a nice web based demo as you suggested, before finalizing the sample app @kartben

The web demo can totally be a follow-up PR (or never happen, that would be fine too :)). We can certainly accept the PR as-is if you simply fix the other editorial comments (sample title, description, ..)

Also, I think I would really like to see you trying to aim for supporting sw0 ... swN, or at least change the current code so that only sw0 sends a message (or, if you really want to keep things as is, the README needs updating I think so that it's more clear that *any* button will work, as right now it's mentioning "user button" in singular form.

Thanks again! 🎹

titouanc commented:
> Also, I think I would really like to see you trying to aim for supporting sw0 ... swN, or at least change the current code so that only sw0 sends a message (or, if you really want to keep things as is, the README needs updating I think so that it's more clear that any button will work, as right now it's mentioning "user button" in singular form.

I believe the best is to accept any input key (in the 0-127 MIDI notes range), so that the demo should work out of the box on any kind of board, even maybe ones with an input keyboard or alike. I adapted the README accordingly

titouanc commented:
Hello @jfischer-no,

Could you please have a look at this ?

jfischer-no commented on samples/subsys/usb/midi/README.rst:
```diff
@@ -0,0 +1,84 @@
+
+.. code-block:: console
+
+  $ amidi -l
```
It does not work, I guess there is something wrong in the interface descriptors.

titouanc replied:
On which board did you try ? On my side, the sample just works out of the box on these 2 boards :

- nucleo_f429zi (https://docs.zephyrproject.org/latest/boards/st/nucleo_f429zi/doc/index.html)
- stm32f4_disco (https://docs.zephyrproject.org/latest/boards/st/stm32f4_disco/doc/index.html)

Both are based on stm32f4, and only support Full-Speed USB; so I haven't been able to test on High-Speed, this may be your case ?

jfischer-no replied:
reel board, bu tit is quite simple function, should work with any board in the tree. I will check again later to see what is wrong.

titouanc replied:
When you retry, could you please also share:

- What host OS are you using (`uname -r`) ?
- What version of Alsa are you using (`amidi --version`) ?
- Are there some useful kernel messages during device enumeration (`dmesg`) ?

In my case:

- `6.11.10-2-MANJARO`
- `amidi version 1.2.13`

```
kernel: usb 1-3.4: new full-speed USB device number 121 using xhci_hcd
kernel: usb 1-3.4: New USB device found, idVendor=2fe3, idProduct=0010, bcdDevice= 4.00
kernel: usb 1-3.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
kernel: usb 1-3.4: Product: USBD MIDI Sample
kernel: usb 1-3.4: Manufacturer: Zephyr Project
kernel: usb 1-3.4: SerialNumber: 31383336303851060041003B
```

titouanc commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,149 @@
+/**
+ * @defgroup usb_midi_cmd MIDI commands
+ * @ingroup usb_midi
+ * @see ump112: 7.3 MIDI 1.0 Channel Voice Messages
```
These are in sub-sections 7.3.1 ... 7.3.7 of "Universal MIDI Packet (UMP) Format and MIDI 2.0 Protocol":
![image](https://github.com/user-attachments/assets/3bd20c6d-b63f-4a89-a520-5e8322eb599f)

titouanc commented on include/zephyr/usb/class/usb_midi.h:
```diff
@@ -0,0 +1,149 @@
+ * @param      p1       The 1st MIDI1 parameter
+ * @param      p2       The 2nd MIDI1 parameter
+ */
+#define UMP_MIDI1(group, command, channel, p1, p2)                                                 \
+	(const uint32_t[])                                                                         \
+	{                                                                                          \
+		(MT_MIDI1_CHANNEL_VOICE << 28) | ((group & 0x0f) << 24) |                          \
+			((command & 0x0f) << 20) | ((channel & 0x0f) << 16) | ((p1 & 0x7f) << 8) | \
+			(p2 & 0x7f)                                                                \
+	}
```
Well, this was my first proposal, but @tmon-nordic suggested to use macros instead, see discussion https://github.com/zephyrproject-rtos/zephyr/pull/81197#discussion_r1839755320 .

Having written both versions (with structs / macros), I have a personal preference for macros. This is because structs require an additional typecast for proper interpretation, for example:

```C
uint32_t packet[4];

if (UMP_MT(packet) == UMP_MT_MIDI1_CHANNEL_VOICE) {
    struct ump_midi1_channel_voice *cv = packet;
    // For example NoteOn / 4 bytes
} else if (UMP_MT(ump) == UMP_MT_DATA) {
    struct ump_data *data = packet;
    // For example SysEx messages / 8 bytes
} else if (UMP_MT(ump) == UMP_MT_SYS_RT_COMMON) {
    struct ump_sys_rt_common *sysrt = packet;
    // For example MIDI clock / 4 bytes
} else if (UMP_MT(ump) == UMP_MT_UMP_STREAM) {
    struct ump_stream *stream = packet;
    // For example MIDI Capability inquiry / 16 bytes
}
```

We may `union` them into a nice top-level `ump` struct, but I feel this is getting complicated, and would yield a bunch of repeated code.

Moreover, some accessor macros can be used for different message types. For example, consider the MIDI1 Channel Voice and System messages that have a similar structure:

![image](https://github.com/user-attachments/assets/3bd20c6d-b63f-4a89-a520-5e8322eb599f)
![image](https://github.com/user-attachments/assets/0696c7c0-20aa-46b8-bdc3-e1741b864e32)

jfischer-no replied:
What I suggested is widely used in Zephyr, e.g. in Bluetooth and networking. You can find many good examples there.
But packet processing in usbd_midi.c should be limited to what is described in "3.2 Data format": Universal MIDI Packet (UMP)", anything else UMP specific should be placed in midi.h (you would need to include in the sample), and maybe in the future there will be a MIDI support that can also be used by e.g. Bluetooth.
As I understand midi class spec, all you need to know before the driver callback is called is the "Message Type". For this you can use a macro defined in usbd_midi.c, if necessary. And then pull the message from the net_buf and pass it to the user driver using a callback like `typedef void (*usb_midi_callback)(const struct device *dev, const uint32_t msg[static 4]);`

titouanc replied:
I splitted the definitions between `midi.h` (Universal MIDI Packet definitions) and `usbd_midi.h` and took the opportunity to improve the doxygen comments. Also changed typing from `const uint32_t *ump` to `const uint32_t ump[static 4]`.

About the macros vs. structs I'm still not convinced (especially considering the number of cases if we were to add more definitions for MIDI2.0 messages; see ump112 section 7.4). Let me know how you think it looks currently; I will rewrite to structs if necessary.

> and maybe in the future there will be a MIDI support that can also be used by e.g. Bluetooth.

Yes, would be awesome to have MIDI support in other "connectivity" stacks :star_struck:

titouanc commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,791 @@
+
+	/* midi20 3.2.2 UMP Messages in a USB Packet: Byte Ordering */
+	words = (uint32_t *)buf->data;
+	for (int i = 0; i < buf->len / 4; i++) {
+		words[i] = sys_le32_to_cpu(words[i]);
+	}
+
+	while (offset < buf->len) {
+		uint32_t *pkt = (uint32_t *)&buf->data[offset];
```
No !

A Universal MIDI Packet can be up to 4 words (of 4 bytes each) long hence the uint32_t pointer passed to the user callback

jfischer-no replied:
Yes, that is how to use net_buf.
Define uint32_t msg[4], pull header, get "Message Type", check type and buf->len, pull the rest of the message, call `typedef void (*usb_midi_callback)(const struct device *dev, const uint32_t msg[static 4]);`

titouanc replied:
Ah then yes, it makes much more sense :+1: 

(I misunderstood your comment as "pass each word by value to the user callback")

titouanc commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,791 @@
+#define USBD_MIDI2_DEFINE_GRPTRM_DESCRIPTOR(inst)                                                  \
+	static const USBD_MIDI2_GRPTRM_DESCRIPTORS_T(                                              \
+		DT_INST_CHILD_NUM(inst)) usbd_midi_grptrm_##inst = {                               \
+		.head =                                                                            \
+			{                                                                          \
+				.bLength = sizeof(struct usbd_midi_grptrm_header_descriptor),      \
+				.bDescriptorType = CS_GR_TRM_BLOCK,                                \
+				.bDescriptorSubtype = GR_TRM_BLOCK_HEADER,                         \
+				.wTotalLength =                                                    \
+					sys_cpu_to_le16(sizeof(USBD_MIDI2_GRPTRM_DESCRIPTORS_T(    \
+						DT_INST_CHILD_NUM(inst)))),                        \
+			},                                                                         \
```
I used `clang-format` with the Zephyr style file. I also found these indentations weird, but I thought this was the intended codestyle.

titouanc commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,791 @@
+	size_t words = UMP_WORDS(ump);
+	size_t buflen = 4 * words;
+
+	LOG_DBG("Send MT=%X group=%X", UMP_MT(ump), UMP_GROUP(ump));
```
Both numbers are 4bits, so only a single character is needed. Not sure if `%1X` would make sense ?

titouanc commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,791 @@
+	size_t buflen = 4 * words;
+
+	LOG_DBG("Send MT=%X group=%X", UMP_MT(ump), UMP_GROUP(ump));
+	if (data->altsetting != ALT_USB_MIDI_2) {
+		LOG_WRN("MIDI2.0 is not enabled");
+		return -EIO;
+	}
```
Thus, you suggest to replace `usbd_midi_set_callback(dev, cb)` with `usbd_midi_set_ops(dev, {iface_ready=..., rx_packet=...})` ?

titouanc replied:
Alternatively, could this be a `usbd_message_type = USBD_MSG_MIDI_REVISION` or something ?

jfischer-no replied:
No, I do not want to abuse it that much, there is a MIDI driver API, if something like that is needed it should be implemented there.

stemschmidt commented:
@titouanc: Thanks for this extension! I build the sample application for a adafruit_feather_nrf52840 board and it worked!

Do you know how to build the midi extension in a C++ project? 

I am having issues with the "static 4" e.g. in 
`int usbd_midi_send(const struct device *dev, const uint32_t ump[static 4]);`

I get this error:
usbd_midi.h:35:65: error: expected primary-expression before 'static'
   35 | int usbd_midi_send(const struct device *dev, const uint32_t ump[static 4]);

fabiobaltieri commented on include/zephyr/usb/class/usbd_midi.h:
```diff
@@ -0,0 +1,51 @@
+ *             -EIO if MIDI2.0 is not enabled by the host
+ *             -EAGAIN if there isn't room in the transmission buffer
+ */
+int usbd_midi_send(const struct device *dev, const uint32_t ump[static 4]);
```
Wow didn't even know the `[static 4]` thing was a thing. So I guess it's C99 (https://www.ibm.com/docs/en/i/7.5?topic=pd-static-array-indices-in-function-parameter-declarations-c-only), I see it's used in other spots in USB drivers which I suppose is fine but if C++ is not having it (in who knows what standards...) I reckon we should avoid it in headers (anything in `include/` really)

Maybe this could be a

```c
struct usbd_midi_ump {
        uint32_t data[4];
}
```
?

titouanc replied:
I wanted to avoid the extra indirection (having to write `ump->data` everywhere), but I guess that would be the least worse compromise here.

Waiting for @jfischer-no re-review to take a decision on this

fabiobaltieri replied:
Yeah ok I understand the intent now but then I really don't think it's a great idea: as designed the API has this logic where it extracts the data length from data packed in the data itself, this I think makes it harder to ensure that it's used correctly, and it exposes some potential out of bound read if data is not correct.

What I think this should do instead is to take a `uint8_t *buf` and `size_t len` as arguments, then the caller can trivially pass the buffer and `sizeof()` of such buffer, the function now instead of extracting the size from the data and assuming that the data is valid can extract the size from the data and compare it with the buffer length passed by the application and validate that the data buffer is big enough. That way the implementation becomes defensive to passing random data, I suppose worst case it would send garbage down the bus but at least it won't read out of buffer or crash as long as the caller does not lie about the buffer length.

titouanc replied:
> this should do instead is to take a uint8_t *buf and size_t len as arguments, then the caller can trivially pass the buffer and sizeof() of such buffer

The UMP spec says that packets are always 1 to 4 uint32, and does not say anything about endianness (which is specified in USB-MIDI2.0 transport), hence why we have an array of uint32. The class driver ensures that only complete packets are delivered to the caller in the USB-MIDI callback. Moreover, the `uint32_t[static 4]` hints the compiler to check that the user always pass a valid memory region, even if only the first item is actually defined (like in MIDI1 packets). Indeed, this may waste up to 12 unused Bytes of memory, but as I see it this comes with better memory safety (also, think of all these `len` that are not passed on the stack :stuck_out_tongue: )

If we were to pass `uint8_t *packet, size_t len` instead, this makes room for inconsistencies. For example: what if `len` is not a multiple of 4 ? What if `len` is not correct compared to the packet Message Type ?

Now, there is still the issue that C++ doesn't understand `uint32_t ump[static 4]`. To solve this, I think that your initial proposal `struct {uint32_t [4]}` in this comment thread makes perfect sense, and I would gladly write it that way.

Because @jfischer-no has also expressed an opinion on these `UMP_*` macros, and that he would have preferred some structs (see https://github.com/zephyrproject-rtos/zephyr/pull/81197/files#diff-6e293d584e7a039998640754f740d3a218bfef304d08b290743ac13da375180b), I just wanted to have a decision on that to finalize the API.

By the way, thank you very much @fabiobaltieri for reviewing !

jfischer-no replied:
This is to be used / and is part of the driver in Zephyr RTOS, we write drivers in C. If C++ cannot handle it, it is the problem of C++ users to work around it.

fabiobaltieri replied:
is it not meant to be called by the application?

stemschmidt replied:
ump is only used in the application interface, in usbd_midi_send and in the usbd_midi_callback.

I do not understand the advantage of using a "typeless" uint32_t pointer instead of a struct which contains a union with the possible payloads? There are two places where a conversion from the bytestream to the struct or the other way around has to be performed which would result in a typed interface... Am I missing something?

fabiobaltieri commented on include/zephyr/usb/class/midi.h:
```diff
@@ -0,0 +1,140 @@
+ * @param[in]  ump    The universal MIDI Packet
+ * @see ump112: 2.1.4 Message Type (MT) Allocation
+ */
+#define UMP_WORDS(ump) (1 + ((0xfe950d40 >> (2 * UMP_MT(ump))) & 3))
```
is there a name for the magic number here? (interesting expression btw, I think it may use a comment)

titouanc replied:
There are 16 message types in UMP, and their length vary between 1 and 4 words; so this is actually an array of 16x2bits numbers packed together as an uint32 :sweat_smile:
In other words, this is not a magic number/constant from the spec, but rather a memory-packed lookup table.

I agree this is worth a comment, will do !

fabiobaltieri replied:
oh, fancy, would it make sense to build the number up in a couple macros with the value and bit shift then? That would probably make it obvious (or at least understandable) without needing a comment in the first place

stemschmidt commented on include/zephyr/usb/class/midi.h:
```diff
@@ -0,0 +1,152 @@
+
+#ifndef ZEPHYR_INCLUDE_USB_CLASS_MIDI_H_
+#define ZEPHYR_INCLUDE_USB_CLASS_MIDI_H_
+
```
Could you please add the 
```
#ifdef __cplusplus
extern "C" {
#endif

```with corresponding
```
#ifdef __cplusplus
}
#endif
```
sections to the header to suppress name mangling in C++? Like in "hid.h"...

titouanc replied:
Unfortunately, this won't work, because the expression `some_array[static 4]` is not valid in C++ (even in an extern "C" section):

```
$ cat example.c 
#ifdef __cplusplus
extern "C" {
#endif

int array_sum(const int arr[static 4])
{
    return arr[0] + arr[1] + arr[2] + arr[3];
}

#ifdef __cplusplus
}
#endif

int main(int argc, const char **argv)
{
    const int array[4] = {1, 2, 3, 4};
    int sum = array_sum(array);
    return 0;
}

$ gcc -std=c99 -o example example.c

$ g++ -o example example.c
example.c:5:29: error: expected primary-expression before ‘static’
    5 | int array_sum(const int arr[static 4])
      |                             ^~~~~~
example.c:5:29: error: expected ‘]’ before ‘static’
    5 | int array_sum(const int arr[static 4])
      |                             ^~~~~~
      |                             ]
example.c:5:29: error: expected ‘)’ before ‘static’
    5 | int array_sum(const int arr[static 4])
      |              ~              ^~~~~~
      |                             )
example.c:5:29: error: expected initializer before ‘static’
    5 | int array_sum(const int arr[static 4])
      |                             ^~~~~~
example.c: In function ‘int main(int, const char**)’:
example.c:17:15: error: ‘array_sum’ was not declared in this scope
   17 |     int sum = array_sum(array);
      |               ^~~~~~~~~

```

stemschmidt replied:
Sorry, I was not clear with my comment, this is an additional request. I know that it would not fix the "static 4" issue. But without 'extern "C"' it is not possible to link it to a C++ project.

stemschmidt replied:
Could you please add the extern "C" to the headers? Otherwise the midi extension cannot be used in a C++ project!

stemschmidt commented on include/zephyr/usb/class/usbd_midi.h:
```diff
@@ -0,0 +1,51 @@
+ */
+
+#include <zephyr/device.h>
+
```
Could you please add the 
```
#ifdef __cplusplus
extern "C" {
#endif

```
with corresponding
```
#ifdef __cplusplus
}
#endif
```
sections to the header to suppress name mangling in C++? Like in "usbd_hid.h"...

rruuaanng commented on dts/bindings/usb/zephyr,usb-midi.yaml:
```diff
@@ -0,0 +1,51 @@
+        First MIDI Group number (address) and number of Group Terminals (size)
+        in this USB-MIDI Group Terminal Block.
+        The MIDI Groups 1 to 16 corresponds to address 0x0 to 0xf. There are at
+        most 16 addressable groups (of 16 channels each) per USB-MIDI interface
```
```suggestion
        most 16 addressable groups (of 16 channels each) per USB-MIDI interface.
```

rruuaanng commented on dts/bindings/usb/zephyr,usb-midi.yaml:
```diff
@@ -0,0 +1,51 @@
+      - "midi1-up-to-128b"
+      - "midi2"
+      description: |
+        Default MIDI protocol of the Group Terminals in this Block
```
```suggestion
        Default MIDI protocol of the Group Terminals in this Block.
```

rruuaanng commented on dts/bindings/usb/zephyr,usb-midi.yaml:
```diff
@@ -0,0 +1,51 @@
+      - "input-only"
+      - "output-only"
+      description: |
+        Type (data direction) of Group Terminals in this Block.
```
You chose to add a period(*`.`*) here, we should add it for the rest of the description, I don't want every sentence to have a different style.

stemschmidt commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,750 @@
+		(struct usb_desc_header *)&usbd_midi_desc_##inst.if1_1_cs_out_ep,        \
+		(struct usb_desc_header *)&usbd_midi_desc_##inst.if1_1_in_ep_fs,         \
+		(struct usb_desc_header *)&usbd_midi_desc_##inst.if1_1_cs_in_ep,         \
+	};                                                                               \
```
Could you please check if a termination of the array is missing, either a NULL or a nil-descriptor (like in e.g. usbd_cdc_acm.c)? init_configuration_inst checks for that condition (https://github.com/zephyrproject-rtos/zephyr/blob/6f240ef18c11473d05aca9954c9ef0d91a19d7db/subsys/usb/device_next/usbd_init.c#L135)

titouanc replied:
Oh that's a very good catch :+1: 

I hadn't noticed that the CDC-ACM nil descriptor was "special", and I was wondering by which magic the end of array was correctly detected by usbd.

stemschmidt replied:
Maybe that caused the issue "It does not work, I guess there is something wrong in the interface descriptors." that @jfischer-no reported?

kartben reviewed (dismissed):
refreshing +1 for docs/sample - thanks!

fabiobaltieri commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,752 @@
+	const struct device *dev = usbd_class_get_private(class_data);
+	const struct usbd_midi_config *cfg = dev->config;
+
+	return (usbd_bus_speed(uds_ctx) == USBD_SPEED_HS)
+		? (direction == USB_EP_DIR_IN)
+			? cfg->desc->if1_1_in_ep_hs.bEndpointAddress
+			: cfg->desc->if1_1_out_ep_hs.bEndpointAddress
+		: (direction == USB_EP_DIR_IN)
+			? cfg->desc->if1_1_in_ep_fs.bEndpointAddress
+			: cfg->desc->if1_1_out_ep_fs.bEndpointAddress;
```
I would split this, have an expilicit if for the outer condition and the inline one for the other... up to you though

fabiobaltieri commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,752 @@
+	};                                                                      \
+	DEVICE_DT_INST_DEFINE(inst, usbd_midi_preinit, NULL,                    \
+			      &usbd_midi_data_##inst, &usbd_midi_config_##inst, \
+			      POST_KERNEL, CONFIG_SERIAL_INIT_PRIORITY, NULL);
```
`CONFIG_SERIAL_INIT_PRIORITY`? you probably wanted to add a specific init priority symbol? or just use `KERNEL_INIT_PRIORITY_DEVICE` I guess

jfischer-no commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,752 @@
+
+static void usbd_midi_rx_work(struct k_work *work)
+{
+	int ret;
+	struct net_buf *buf;
+	struct usbd_midi_data *data = CONTAINER_OF(work, struct usbd_midi_data, rx_work);
```
Please order them as an upside down tree, following the style in other USB code.

jfischer-no commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,752 @@
+	return buf;
+}
+
+static inline uint8_t usbd_midi_get_ep(struct usbd_class_data *const class_data,
+				       uint8_t direction)
+{
+	struct usbd_context *uds_ctx = usbd_class_get_ctx(class_data);
+	const struct device *dev = usbd_class_get_private(class_data);
+	const struct usbd_midi_config *cfg = dev->config;
+
+	return (usbd_bus_speed(uds_ctx) == USBD_SPEED_HS)
+		? (direction == USB_EP_DIR_IN)
+			? cfg->desc->if1_1_in_ep_hs.bEndpointAddress
+			: cfg->desc->if1_1_out_ep_hs.bEndpointAddress
+		: (direction == USB_EP_DIR_IN)
+			? cfg->desc->if1_1_in_ep_fs.bEndpointAddress
+			: cfg->desc->if1_1_out_ep_fs.bEndpointAddress;
+}
```
Must be like https://github.com/zephyrproject-rtos/zephyr/blob/2a04abafa0241e77956d11c5df0407e3b3fe7aec/subsys/usb/device_next/class/usbd_cdc_ncm.c#L245-L271

This construction offers literally no advantages other than unreadability. Please let the compiler to be smart ass. In general, the code should follow the style of existing class/function implementations.

jfischer-no commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,752 @@
+		},                                                                       \
+		.if0_cs = {                                                              \
+			.bLength = sizeof(struct usb_midi_cs_ac_header_descriptor),      \
+			.bDescriptorType = CS_INTERFACE,                                 \
```
There are USB_DESC_CS_INTERFACE/USB_DESC_CS_ENDPOINT, they should be used instead.

titouanc commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,752 @@
+	struct usb_if_descriptor if0_std;
+	struct usb_midi_cs_ac_header_descriptor if0_cs;
+
+	/* Empty MidiStreaming 1.0 on altsetting 0 */
+	struct usb_if_descriptor if1_0_std;
+	struct usb_midi_header_descriptor if1_0_ms_header;
+	struct usb_ep_descriptor if1_0_out_ep_fs;
+	struct usb_ep_descriptor if1_0_out_ep_hs;
+	struct usb_midi_cs_endpoint_descriptor if1_0_cs_out_ep;
+	struct usb_ep_descriptor if1_0_in_ep_fs;
+	struct usb_ep_descriptor if1_0_in_ep_hs;
+	struct usb_midi_cs_endpoint_descriptor if1_0_cs_in_ep;
```
I don't understand why this wouldn't work ?

From what I understand, Linux attempts to probe a USB-MIDI2 interface and falls back to USB-MIDI1 if such a valid interface does not exit (https://github.com/torvalds/linux/blob/v6.12/sound/usb/midi2.c#L1075-L1105). Moreover, I can tell that the sample from this PR is working with both Linux & Mac OS hosts for me. Other reviewers reported that the USB-MIDI stack works for them too.

From the USB-MIDI2 spec, 3.1.1 MIDI Streaming Interface with Two Alternate Settings: Backward Compatibility:

> The first MIDI Streaming Interface exposed by the Device should contain an Alternate
> Setting 0 which is compliant with USB Device Class Specification for MIDI Devices
> Version 1.0. In other words, if Host software selects the first indexed Alternate Setting on
> the MIDI Streaming Interface, the Device should expose a MIDI Function that is compliant
> with USB MIDI 1.0. The bcdMSC field in the Class-Specific MIDI Streaming Interface
> Header shall be set to 0x0100. This requirement allows the MIDI Function to interoperate
> with existing Hosts exactly the same as current legacy MIDI 1.0 Functions.
> The first MIDI Streaming Interface exposed by the Device should contain an Alternate
> Setting 1 which contains the preferred device implementation which conforms to the new
> design defined by this USB MIDI 2.0 specification. The bcdMSC field in the Class-Specific
> MIDI Streaming Interface Header shall be set to 0x0200. Further Alternate Settings which
> conform to the new design defined by this USB MIDI 2.0 specification may be included in
> Alternate Setting 2 or higher.

Nothing says that these 2 altsettings should describe the same topology (and actually they couldn't because elements do not exist in USB-MIDI2, or MIDI2 only groups cannot be represented in USB-MIDI1). Therefore in this implementation, the altsetting 0 contains a perfectly valid empty Midistreaming interface, that doesn't contain any jack bound to its endpoints. Therefore, there is no MIDI port for the host to communicate through, hence no need to implement anything further.

I do see value in a fully fledged USB-MIDI1 implementation, bound to the altsetting 0. However i see this as an extra feature, because it requires an additional layer to translate UMPs into the "old" usb-midi1 format and the other way around. I will gladly do this in a following PR.

tmon-nordic replied:
> I did not find anything in the specification that says you can skip the MIDI1 interface implementation. Rather, you are required to provide a valid interface implementation for backward compatibility.

It is not required. It seems that you skipped the "1.5 Reserved Words and Specification Conformance" the word "should" is reserved for "Statements of recommendation" which are "Recommended but not mandatory. An implementation that does not conform to some or all ‘should’ statements is still conformant, providing all ’shall’ statements are conformed to."

All the MIDI1 references are with "should" and therefore the compliant implementation can skip it.

jfischer-no replied:
Yes, I missed the "shall" definition. Then I doubt the need to have the MIPI1 interface, it is not required by the spec, but then the Linux implementation would not recognize the MIPI2 interface, so I am fine as it is. I also see that you have updated the interface descriptors, which makes them less confusing. I also find out why it is not detected on some of my machines with different and newer kernel versions, someone had the bad idea not to set the CONFIG_SND_USB_AUDIO_MIDI_V2 option in the Debian kernel.

titouanc commented:
Hello again @jfischer-no, and thank you for your latest review !

I've once again addressed your (and other reviewers) comments, and rebased on top of `main`. In addition, I updated the API from `usbd_midi_set_callback(dev, cb)` to `usbd_midi_set_ops(dev, {rx_packet_cb=..., status_change_cb=...})`, as discussed here https://github.com/zephyrproject-rtos/zephyr/pull/81197#discussion_r1880528810. This allows the application to know if the USB-MIDI2.0 interface is enabled or not. I updated the sample accordingly (to light up a LED when USB-MIDI2.0 is enabled).

There are still a couple of open points, for which I would really like to have a clear answer:

- `UMP_*` macros vs. struct definitions (https://github.com/zephyrproject-rtos/zephyr/pull/81197#discussion_r1880170743). Do you agree with the current macros based proposal ?
- Have you managed to successfully run the sample (https://github.com/zephyrproject-rtos/zephyr/pull/81197#discussion_r1906157780) ? Once again, I would be glad to add full support for USB-MIDI1.0, but in a second PR (as this looks like an additional feature to me). In the meantime, do you agree that the current implementation is mergeable ?

titouanc commented on doc/connectivity/usb/device_next/api/index.rst:
```diff
@@ -11,3 +11,4 @@ New USB device support APIs
    usbd_hid_device.rst
    uac2_device.rst
    usbd_msc_device.rst
+   usb_midi.rst
```
Because I initially had one commit, and because Github uses the commit message by default for the PR description, this made for a nice description.

Will rewrite the commit message with the bare links only :+1:

titouanc commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,787 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbd_midi, CONFIG_USBD_MIDI_LOG_LEVEL);
+
+#define ALT_USB_MIDI_1 0x00
+#define ALT_USB_MIDI_2 0x01
```
I find it more readable to have a name in the API implementation, especially because alt0=MIDI1 and alt1=MIDI2 (off-by-one; easy to confuse 1 for MIDI1), but this is definitely a matter of taste.

I have no opinion on ALT_USB_MIDI_1 vs. MIDI1_ALTERNATE, so I will apply your suggestion

diegoherranz commented on doc/connectivity/usb/device_next/api/usb_midi.rst:
```diff
@@ -0,0 +1,12 @@
+.. _usb_midi:
+
+MIDI 2.0 Class device API
+#########################
+
+USB MIDI 2.0 device specific API defined in :zephyr_file:`include/zephyr/usb/class/usbd_midi.h`.
+
+API Reference
+*************
+
+.. doxygengroup:: usb_midi
+.. doxygengroup:: usb_midi_ump
```
I think a similar reasoning can be applied to the recently renamed include/zephyr/audio/midi.h. There are many references to USB there, which I think could be removed since that could be used for transports other than USB.

jfischer-no commented on include/zephyr/usb/class/usbd_midi.h:
```diff
@@ -0,0 +1,80 @@
+ *             -EIO if MIDI2.0 is not enabled by the host
+ *             -EAGAIN if there isn't room in the transmission buffer
+ */
+int usbd_midi_send(const struct device *dev, const struct midi_ump *ump);
```
https://github.com/zephyrproject-rtos/zephyr/pull/81197#discussion_r1925732707

titouanc replied:
Sorry to over-write on this, but I just want to make sure that I fully understand your intent. When you say

> what you call a sample is half of the driver.

this is because `usbd_midi_send` is not a syscall, and therefore the caller is not isolated from kernelspace ? Is there any other reason ?

Like I said, I prefer the simplicity of `uint32_t [static 4]`, but this is functionally equivalent to `struct {uint32_t [4]}`, and they have the same memory layout. Besides esthetic considerations, is there any other advantage of the former over the struct form ?

In my project using usb-midi, I implement UMP Stream (discovery and dynamic function block configuration). I do the following, which just works fine (in plain C), so I don't need to bother with structs at all:

```C
uint32_t reply[4];
/* build reply */
usbd_midi_send(dev, (const struct midi_ump *) reply);
```

If we are to go for the static array form, how are C++ users supposed to do ? If I understand correctly, they would need to add a some boilerplate (to compile in C) that understands `uint32_t [static 4]` and expose a C++ friendly interface, which sounds like an extra burden without clear advantage to me. Do you have any other suggestion ?

Thanks again for the review, and sorry for the lengthy discussion.

jfischer-no replied:
No, because the sample is actualy the driver. You are using the Zephyr driver model and DT to implement a MIDI driver with a USB interface. In Zephyr, the drivers are written in C. If someone for whatever reason decides to write them in C++ or some other language in downstream work, it is their responsibility to provide the proper interface. I do not see why the code here should suffer if someone uses a different language. There are no real dependencies on UMP in the USB MIDI2 spec, other than determining the length. The macro to get the length can be defined in usbd_midi.c, and there would be no need to include midi.h. Together with `int usbd_midi_send(const struct device *dev, const uint32_t ump[static 4]);`, this implementation would provide a tiny, clean and safe interface, independent of any future changes to midi.h.

stemschmidt replied:
Feel free to correct me if I am on a complete wrong track, but an application will send midi data with usbd_midi_send. Therefore it has to include include/zephyr/usb/class/usbd_midi.h

According to this information from the zephyr doc page (https://docs.zephyrproject.org/latest/develop/languages/cpp/index.html#header-files-and-incompatibilities-between-c-and-c) some care has to be taken to allow headers to be included from C and C++.

If "uint32_t [static 4]" is not available in C++ it should not appear in a header file that has to be included in order to use a certain functionality. This is not writing a driver in another language, this is simply ensuring that the driver can be used from C and C++.

jfischer-no replied:
@stemschmidt It might help if you read my comments where I said that it is still the driver that calls usbd_midi_send(). That is the design behind using the Zephyr device model and devicetree. Limitations of the language you use must not weaken the upstream code.

stemschmidt replied:
I read it, but I did not understand it. :-) And still do not understand it: How should an application send MIDI messages?

jfischer-no commented on include/zephyr/audio/midi.h:
```diff
@@ -0,0 +1,204 @@
+/**
+ * @brief      Universal MIDI Packet container
+ */
+struct midi_ump {
+	uint32_t data[4]; /**< Raw content */
+};
```
@titouanc Not sure how other interfaces would require it, e.g. Bluetooth. You could change usbd_midi_send() to int `usbd_midi_send(const struct device *dev, const struct midi_ump ump);`, so it would be more safe, and then I would also be happy.

Also, please add a comment that the variables must be in the CPU native endianness.

jfischer-no commented on include/zephyr/audio/midi.h:
```diff
@@ -0,0 +1,204 @@
+
+/**
+ * @brief Universal MIDI Packet definitions
+ * @defgroup usb_midi_ump USB MIDI 2.0 Universal MIDI Packet definitions
```
Please make it interface neutral, remove all the `usb_` prefixes and `USB`.

titouanc commented:
Thanks again for you latest review @jfischer-no 

I just applied your comments, and rebased on top of main:
- Rename `ALT_USB_MIDI_x` into `MIDIx_ALTERNATE`
- Pass `ump` by value (instead of by pointer) in `usbd_midi_send` and `rx_packet_cb`. Also adapted the macros in `midi.h` to take a struct (and not a pointer) as parameter for consistency
- Remove all references to USB in `midi.h` and attach it to the audio group in Doxygen
- Rewrite the first commit message to remove Markdown-like syntax

jfischer-no commented on include/zephyr/usb/class/usbd_midi.h:
```diff
@@ -0,0 +1,80 @@
+
+/**
+ * @brief      Send a Universal MIDI Packet to the host
+ * @param[in]  dev   The USB-MIDI interface
```
` * @param[in]  dev   MIDI2 device`

If you want to pass an interface, then you must not use Zephyr device model and DT, might be a better approach or not, if you are curious, check out the new USB DFU implementation PR.

jfischer-no commented on subsys/usb/device_next/CMakeLists.txt:
```diff
@@ -71,6 +71,11 @@ zephyr_library_sources_ifdef(
 )
 
+zephyr_library_sources_ifdef(
+  CONFIG_USBD_MIDI_CLASS
```
`  CONFIG_USBD_MIDI2_CLASS`

jfischer-no commented on subsys/usb/device_next/CMakeLists.txt:
```diff
@@ -71,6 +71,11 @@ zephyr_library_sources_ifdef(
 
+zephyr_library_sources_ifdef(
+  CONFIG_USBD_MIDI_CLASS
+  class/usbd_midi.c
```
`  class/usbd_midi2.c`

Sorry, I missed that, but it needs to be fixed.

jfischer-no commented on subsys/usb/device_next/class/Kconfig:
```diff
@@ -10,3 +10,4 @@ rsource "Kconfig.bt"
 rsource "Kconfig.msc"
 rsource "Kconfig.uac2"
 rsource "Kconfig.hid"
+rsource "Kconfig.midi"
```
`rsource "Kconfig.midi2"`

jfischer-no commented on subsys/usb/device_next/class/Kconfig.midi:
```diff
@@ -0,0 +1,17 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config USBD_MIDI_CLASS
```
s/USBD_MIDI_CLASS/USBD_MIDI2_CLASS

jfischer-no commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,786 @@
+#include <zephyr/usb/class/usbd_midi.h>
+#include <zephyr/usb/usbd.h>
+
+#include "usbd_uac2_macros.h"
```
Do you still need to include it?

titouanc replied:
Yes, there are a handful of defines that are reused from there (to name a few):
* `AUDIO` class code https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device_next/class/usbd_uac2_macros.h#L51
* `MIDISTREAMING` subclass code https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device_next/class/usbd_uac2_macros.h#L57
* `AUDIOCONTROL` subclass code https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device_next/class/usbd_uac2_macros.h#L55

jfischer-no commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,786 @@
+#include "usbd_uac2_macros.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbd_midi, CONFIG_USBD_MIDI_LOG_LEVEL);
```
LOG_MODULE_REGISTER(usbd_midi2, CONFIG_USBD_MIDI2_LOG_LEVEL);

jfischer-no commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,786 @@
+	struct usbd_midi_data *data = dev->data;
+	struct udc_buf_info *info = udc_get_buf_info(buf);
+
+	LOG_DBG("USB-MIDI request for %s ep=%d len=%d err=%d", dev->name, info->ep, buf->len, err);
```
Endpoints should always be logged as `ep 0x%02x`
```c
	LOG_DBG("MIDI2 request for %s ep 0x%02x len %u err %d",
		dev->name, info->ep, buf->len, err);
```

jfischer-no commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,786 @@
+		data->ops.status_change_cb(dev, USBD_MIDI_2_ENABLED);
+	}
+
+	LOG_DBG("USB-MIDI enable for %s", dev->name);
```
Prefixing logging message with USB-MIDI does not make any sense, they are already prefixed with `usbd_midi2:`

LOG_DBG("Enable %s", dev->name);

jfischer-no commented on subsys/usb/device_next/class/usbd_midi.c:
```diff
@@ -0,0 +1,786 @@
+	+ DT_INST_CHILD_NUM_STATUS_OKAY(inst)             \
+	* sizeof(struct usb_midi_grptrm_block_descriptor)
+
+#define USBD_MIDI_DEFINE_DESCRIPTORS(inst)                                               \
```
I prefer just `n` and want it to be consistent with other class implementations `#define USBD_MIDI_DEFINE_DESCRIPTORS(n)`

jfischer-no reviewed (changes requested):
There are a few inconsistencies, please fix them.

titouanc commented on dts/bindings/usb/zephyr,usb-midi.yaml:
```diff
@@ -0,0 +1,51 @@
+
+child-binding:
+  description: |
+    USB MIDI Group terminal block.
```
Mmmh, not really. The concept of Group Terminal (Blocks) is in the USB-MIDI specification, not in MIDI2 (UMP)

jfischer-no replied:
https://github.com/zephyrproject-rtos/zephyr/pull/81197#discussion_r1935337348

titouanc commented on dts/bindings/usb/zephyr,usb-midi.yaml:
```diff
@@ -0,0 +1,51 @@
+# Copyright (c) 2024 Titouan Christophe
+# SPDX-License-Identifier: Apache-2.0
+
+description: USB MIDI Class
```
I find this confusing. We can exchange MIDI1 over a USB-MIDI2 device

titouanc replied:
I would agree with `USB-MIDI2 device` though

jfischer-no replied:
This YAML binding is for MIDI devices following specification release 2.0. So it should be "MIDI2 device" and use MIDI2 consistently. There is not need to place hyphen in the description between USB and MIDI2. (be consistent with USB DFU, USB HID, USB CDC ACM....).

titouanc commented:
Hello @jfischer-no ! I just applied your comments and rebased on top of main:

- rename `(usbd_)midi -> midi2`
- Improve consistency of comments/documentation relative to USB-MIDI/USB_MIDI/MIDI2 etc...
- Change `ops.status_change_cvb(dev, enum status) -> ops.ready_cb(dev, const bool)`
- Use `n` instead of `inst` in DT macros
- Clean up LOG_DBG messages to remove unnecessary `USB-MIDI` prefix
- Rename DT bindings
- Return `-ENOBUFS` instead of `-EAGAIN` in usbd_midi_send if there's enough space in the Tx buffer

jfischer-no reviewed (approved):
LGTM, thank you @titouanc, happy FOSDEM time.

titouanc commented:
Thanks a lot for getting this ready just on time @jfischer-no :raised_hands:

kartben reviewed (approved):
w00t! Thanks @jfischer-no for your help!

kartben commented on include/zephyr/usb/class/usbd_midi.h:
```diff
@@ -0,0 +1,80 @@
+ * @param[in]  ump   The packet to send, in Universal MIDI Packet format
+ * @return     0 on success
+ *             -EIO if MIDI2.0 is not enabled by the host
+ *             -EAGAIN if there isn't room in the transmission buffer
```
was left in a pending review of mine, apparently -- will be good to address as a follow-up

diegoherranz commented:
Many thanks for your work here, guys. Really keen on giving this a try.

ho-ho-ho commented:
I don't know if I should open a new issue for that or not, but I've just managed to test it under arch linux (works flawlessly) and Windows 10 (22H2) ... and I had to find out that windows does not have a MIDI 2.0 driver yet, so it doesn't work at all. It looks like they are in the process adding it [with a preview version out last week](https://devblogs.microsoft.com/windows-music-dev/announcing-windows-midi-services-customer-preview-1/), but until that goes live this driver won't work on windows :(
